### DISCUSSION TITLE
Checkout error in container runner

### DISCUSSION THREAD

**Discussion by aleks-ivanov** [community/community#26811] [Actions]
<p>This is the non-root user I am creating in the Docker image to run the workflow:</p>
<pre><code class="lang-auto">RUN useradd --shell /bin/bash --create-home cicd-runner
USER cicd-runner
</code></pre>
<p>When I try to simply checkout the repo with the official checkout action, this error appears:</p>
<pre><code class="lang-auto">  /usr/bin/git init /__w/pydeps/pydeps
  /__w/pydeps/pydeps/.git: Permission denied
  Error: The process '/usr/bin/git' failed with exit code 1
</code></pre>
<p>What permissions do I need to give the non-root user ?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>There are no permissions you can set inside the container image to make this work. The workspace <em>on the runner VM</em> is mounted to <code>/__w/</code> when the runner starts the job container (check the “Initialize containers” step in the log). Assuming a GitHub hosted runner VM the workspace is owned by the user <code>runner</code>. So what’s happening here is that the container is running as a <em>different</em> non-root user and can’t write another user’s files.</p>
<p>The <a href="https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions#user">documentation on Dockerfiles for Actions</a> mentions the following regarding <code>USER</code> instructions:</p>
<blockquote>
<p>Docker actions must be run by the default Docker user (root). Do not use the <code>USER</code> instruction in your <code>Dockerfile</code>, because you won’t be able to access the <code>GITHUB_WORKSPACE</code>.</p>
</blockquote>
<p>If you use a job container you can’t adjust ownership before passing control to the container, so I’m afraid you’ll have to either run the container as root or start it manually (<code>docker run</code> and possibly <code>docker exec</code>).</p>

**Comment by aleks-ivanov** [comment]
<p>Yeah, that’s what I was afraid of… thank you for the clarification :slightly_smiling_face:</p>

**Comment by aleks-ivanov** [comment]
<p>Just to confirm if I understood correcly, does that mean that the GiHA runtime is installing a non-root user (<code>runner</code>) under the hood and ensuring the upcoming GiHA actions are executed within the context of this non-root (<code>runner</code>) user (installed by the GiHA runtime) ?</p>

**Reply by n1ngu** [reply]
Not quite but sort of. GHA ubuntu VMs run with a UID=1001 user and the runner workspace is owned by that and mounted as is in the docker container. If your container runs as non-root, non-1001, you won't own the workspace :smiling_face_with_tear: .

You can run the container with `options: --user=1001` even if that UID did not exist inside the container userspace, most containers will work because a home and a workspace are mounted belonging to 1001. You can try it at home with some container e.g. `docker run --rm -ti --user=1234 debian`: your uid does not have a named user but everything "works".

But if your container featured an environment and file system tailored for the non-root, non-1001 user, then you are mostly screwed.

**Comment by airtower-luna** [comment]
<p>Not sure what you mean by “under the hood”. On a GitHub hosted runner the <code>runner</code> user is part of the standard VM setup, not sure about self-hosted ones.</p>
<p>Actions stuff that runs on the VM (<code>run</code> without a job container, Javascript Actions) runs as the <code>runner</code> user. You can use <code>sudo</code> without password for commands that need root access. Job containers and Docker Actions run as the default user for the image, which is <code>root</code> unless there’s a <code>USER</code> instruction.</p>
<p>The workspace directory in the VM is owned by the <code>runner</code> user. <code>root</code> naturally can write there anyway, but any other user defined in a container that doesn’t happen to have the exact same UID as the <code>runner</code> user won’t be able to.</p>

**Comment by n1ngu** [comment]
Related https://github.com/actions/checkout/issues/47
